torch/lib/c10d/ProcessGroupWrapper.cpp

```diff
@@ -23,46 +23,68 @@ struct CollectiveFingerPrint {
   // Ref to input tensors, if given, of the collective. If given, shapes will be
   // checked across processes to ensure valid input into the collective.
   const std::vector<at::Tensor>& input_tensors_;
+  // input tensor data types
+  std::vector<int> tensor_dtypes_;
+  // input tensor device types
+  std::vector<int8_t> tensor_device_types_;
   explicit CollectiveFingerPrint(
       OpType op_type,
       const std::vector<at::Tensor>& input_tensors)
-      : op_type_(op_type), input_tensors_(input_tensors) {}
-
-  // Verifies a given int is the same across processes.
-  void verify_num(
-      int64_t value,
-      const c10::intrusive_ptr<ProcessGroup>& pg,
-      const std::string& failureMsg) {
-    auto tensor = at::full({1}, value, at::TensorOptions().dtype(at::kLong));
-    std::vector<at::Tensor> tensors;
-    tensors.reserve(pg->getSize());
-    for (int i = 0; i < pg->getSize(); ++i) {
-      tensors.emplace_back(at::zeros_like(tensor));
+      : op_type_(op_type), input_tensors_(input_tensors) {
+    tensor_dtypes_.reserve(input_tensors.size());
+    tensor_device_types_.reserve(input_tensors.size());
+    for (const at::Tensor& t : input_tensors_) {
+      tensor_dtypes_.push_back(static_cast<int8_t>(t.dtype().toScalarType()));
+      tensor_device_types_.push_back(static_cast<int8_t>(t.device().type()));
     }
-    std::vector<std::vector<at::Tensor>> out_tensors({tensors});
-    std::vector<at::Tensor> inp_tensors({tensor});
-    pg->allgather(out_tensors, inp_tensors)->wait();
-    std::unordered_set<int64_t> gathered;
-    for (const auto& t : out_tensors[0]) {
-      auto n = t.item().to<int64_t>();
-      gathered.insert(n);
-      if (gathered.size() > 1) {
-        TORCH_CHECK(false, failureMsg);
+  }
+
+  // Logs collective information in case of a failure.
+  friend std::ostream& operator<<(
+      std::ostream& output,
+      const CollectiveFingerPrint& collective_fingerprint);
+
+  at::Tensor serialize_fingerprint() {
+    std::vector<int64_t> data;
+    // OpType
+    data.push_back(static_cast<int64_t>(op_type_));
+    // Shapes
+    for (const auto& tensor : input_tensors_) {
+      auto sizes = tensor.sizes().vec();
+      for (const auto& s : sizes) {
+        data.push_back(s);
       }
     }
+    // tensor dtypes
+    for (const auto& type : tensor_dtypes_) {
+      data.push_back(type);
+    }
+    // device types
+    for (const auto& d : tensor_device_types_) {
+      data.push_back(d);
+    }
+    // Serialize data into tensor
+    int64_t data_size = data.size();
+    // Need to clone here otherwise the data memory is not copied and would be
+    // released after this function returns.
+    at::Tensor serialized_tensor =
```

Comment:
Not sure how often we call `serialize_fingerprint()`, but if it is in a performance critical code path, we can avoid cloning here, by directly constructing a `c10::Storage` instance of a `std::vector<>`.

Reply:
It actually is not on a perf critical codepath (actually this path inserts collectives before the real ones as a debug check, so it is intentionally not for performance and more for application debugging), but I'll see if I can switch this over to c10::Storage which is a better convention in the codebase.

Reply:
I tried switching this over to c10::Storage, but wasn't able to remove the clone (still getting garbage data if I remove it). Does anything stand out as glaringly wrong?